doc/developer/design/20230528_managed_clusters.md

```diff
@@ -0,0 +1,354 @@
+ALTER CLUSTER foo REPLICATION FACTOR 0;
+
+-- You can also create a cluster with multiple replicas from the get go.
+CREATE CLUSTER foo2 SIZE 'small', REPLICATION FACTOR 2;
+```
+
+## Detailed description
+
+### Managed clusters
+
+We'll introduce the concept of a **managed** cluster. A managed cluster is one
+with a declared size and replication factor, where Materialize is responsible
+for ensuring the replica set matches the declared size and replication factor.
+The replicas of a managed cluster are visible in the system catalog, but cannot
+be directly modified by users.
+
+Existing clusters, where users control the replica set manually via
+`{CREATE|DROP} CLUSTER REPLICA`, will be deemed **unmanaged** clusters.
+
+### `CREATE CLUSTER`
+
+The `CREATE CLUSTER` statement will learn three new options:
+
+  * `SIZE`, a string, which specifies the desired size of the replicas in the
+    cluster.
+  * `MANAGED`, a boolean, which specifies whether the cluster is managed or
+    unmanaged.
+  * `REPLICATION FACTOR`, an integer, which specifies the desired number of
+    replicas.
+
+Users must specify either the `SIZE` or `REPLICAS` option when creating a
+cluster. The two options may not be specified simultaneously.
+
+The `MANAGED` option must be true if `SIZE` is specified or false if `REPLICAS`
+is specified. It takes on the appropriate default if unspecified. We expect that
+this option will never be explicitly specified in practice, but we accept it in
+order to be consistent with `ALTER CLUSTER`.
+
+The `REPLICATION FACTOR` option may only be specified if `MANAGED` is true. If unspecified, it defaults to `1`.
+
+For managed clusters, Materialize will automatically create replicas named
+`r1`, `r2`, ..., `rN`, where `N` is the desired replication factor, all of
+the size specified by the `SIZE` option.
+
+For unmanaged clusters, `CREATE CLUSTER`'s behavior remains unchanged.
+
+### `ALTER CLUSTER`
+
+#### Changing size and replication factor
+
+We'll extend the `ALTER CLUSTER` statement to support altering the `SIZE` and
+`REPLICATION FACTOR` of a cluster using the usual syntax for `ALTER ... SET`
+commands:
+
+```sql
+-- Change the size of a cluster to `medium`.
+ALTER CLUSTER c SET (SIZE = 'medium');
+-- Change the replication factor of a cluster to 2.
+ALTER CLUSTER c SET (REPLICATION FACTOR = 2);
+-- Simultaneously change the size and replication factor of a cluster.
+ALTER CLUSTER c SET (SIZE = 'medium', REPLICATION FACTOR = 2);
+```
+
+When the command updates the cluster's size, Materialize will immediately drop
+all existing replicas of the cluster and recreate them with the same names with
+the new size.
+
+> **Warning**
+>
+> In the initial implementation, changing a cluster's size will result in
+> downtime while the new replicas rehydrate. This may be surprising to users.
+>
+> We hope to change the behavior in a future release so that no downtime is
+> incurred, but this will require [dynamic cluster scheduling].
```

Comment:
I'm not convinced it requires the full cluster scheduling infrastructure. If we had a rehydration signal, we could leave the old replicas around until the new ones are hydrated, minimizing downtime as a result. This would not require a complete controller+scaling policy. As a stop-gap, we could even offer to delay tearing down the old replicas by a fixed amount of time.

Reply:
> If we had a rehydration signal

Isn't this the hard part, though? It's not immediately clear to me how we'd build such a rehydration signal. I've always imagined that you need some sort of cluster scheduling widget (could just be the existing controller, idk) that monitors the uppers of each index in each replica and waits until the uppers are "about the same" to declare success.

> As a stop-gap, we could even offer to delay tearing down the old replicas by a fixed amount of time.

I do like this stopgap.